File: flink-code-scala-0421/src/main/scala/com/atguigu/day02/util/SensorSource.scala
package com.atguigu.day02.util

import java.util.Calendar

import org.apache.flink.streaming.api.functions.source.SourceFunction.SourceContext
import org.apache.flink.streaming.api.functions.source.{RichParallelSourceFunction, SourceFunction}

import scala.util.Random

class SensorSource extends RichParallelSourceFunction[SensorReading] {

  var running = true

  override def run(sourceContext: SourceContext[SensorReading]): Unit = {
    val rand = new Random

    var curFTemp = (1 to 10).map(
      i => ("sensor_" + i, rand.nextGaussian() * 20)
    )

    while (running) {
      curFTemp = curFTemp.map(
        t => (t._1, t._2 + rand.nextGaussian() * 0.5)
      )
      val curTime = Calendar.getInstance.getTimeInMillis

      curFTemp.foreach(t => sourceContext.collect(SensorReading(t._1, curTime, t._2)))

      Thread.sleep(100)
    }
  }

  override def cancel(): Unit = running = false
}

File: flink-code-java-0421/src/main/java/com/atguigu/day02/util/SensorSource.java
package com.atguigu.day02.util;

import org.apache.flink.streaming.api.functions.source.RichParallelSourceFunction;

import java.util.Calendar;
import java.util.Random;

public class SensorSource extends RichParallelSourceFunction<SensorReading> {

    private boolean running = true;

    @Override
    public void run(SourceContext<SensorReading> sourceContext) throws Exception {
        Random rand = new Random();

        String[] sensorIds = new String[10];
        double[] curFTemp = new double[10];
        for (int i = 0; i < 10; i++) {
            sensorIds[i] = "sensor_" + (i + 1);
            curFTemp[i] = 65 + (rand.nextGaussian() * 20);
        }
        while (running) {
            long curTime = Calendar.getInstance().getTimeInMillis();
            for (int i = 0; i < 10; i++) {
                curFTemp[i] += rand.nextGaussian() * 0.5;
                sourceContext.collect(new SensorReading(sensorIds[i], curTime, curFTemp[i]));
            }
            Thread.sleep(100);
        }
    }

    @Override
    public void cancel() {
        this.running = false;
    }
}
